client/src/optimiser/Optimiser.js

```diff
@@ -152,28 +152,49 @@ class Optimiser {
     // Stop performance tracking
     const t1 = performance.now();
     const time = t1 - t0;
-    console.log(timetables);
+    // console.log(timetables); //running this will be resource heavy
+    // https://stackoverflow.com/questions/11426185/will-console-log-reduce-javascript-execution-performance
 
     return {timetables, time};
   }
 
-  applyTimeRestrictions(earliestStart, latestFinish) {
+  //we just need to change this to do each subject individually
+  applyTimeAndDeliveryRestrictions(earliestStart, latestFinish, deliveryPref) {
     /**
      * General Methodology:
-     * The idea is to apply the restrictions as requested, but if the
+     * The idea is to apply the restrictions as requested, and if the
      * restrictions end up with a subject completely missing a classtype,
-     * such as a Tutorial 1 - the restrictions are too tight.
+     * such as a Tutorial 1 - add a violating class in order to
+     * have a functioning timetable.
```

Comment:
Sorry forgot to mention: The reason why I am doing this is specifically for the feature. If you don't make this change then the delivery mode filter won't work well. For example if you have 3 subjects that are in-person and 1 subject that is online only, then `applyRestrictions` won't be able to generate any timetables that are in-person only, because all of the online-only subjects will be removed, causing it to return false. Vice-versa.

Reply:
Ah right. Do you know why a student would ever have see how a student would ever have a mixture of online-only and on campus subjects and choose the "online" delivery _preference_? I emphasise preference because there's a different between a preference (sorting criteria) and a restriction (class removal criteron).

Reply:
I agree with you that there would be little reason for a student to
preference online -- if they were overseas they would want to restrict
completely to online only.

But I could imagine a few scenarios: for example a student who lives far
away from campus and wants to minimise the amount they need to travel to
campus.

However (I think you know this just confirming :) ) I wasn't modifying the
feature from the online only perspective, but rather the in-person only
perspective, where most people would want to preference being in person,
rather than restrict to in person only.

Also consider that prefering over restricting only ever happens after the
algorithm has exhausted it's completely perfect options, (and in the old
case it   gives out a random timetable, ignoring restrictions) so I don't
think there's ever a scenario where a user would want to be given no
options if their restrictions are too tight - but they probably would want
to be notified. So I don't see any downside to having it be preference based as opposed to restrictive.

Reply:
I see, these are valid points. Happy to go forward with this :)